server/src/main/java/org/opensearch/index/engine/Engine.java

```diff
@@ -1291,6 +1298,20 @@ public void failEngine(String reason, @Nullable Exception failure) {
                             );
                         }
                     }
+
+                    // If cleanup unreferenced flag is enabled and force merge or regular merge failed due to disk full,
+                    // cleanup all unreferenced files created during failed and reset the shard state back to last
+                    // Lucene Commit.
+                    if (shouldCleanupUnreferencedFile()
+                        && (reason.equals("force merge") || reason.equals("merge failed"))
+                        && failure != null
+                        && failure.getCause() != null
+                        && failure.getCause().getCause() != null
+                        && failure.getCause().getCause().getMessage() != null
+                        && failure.getCause().getCause().getMessage().contains("No space left on device")) {
```

Comment:
Should we extract this as a helper? This will make the `if` check more readable here

Reply:
Addressed the comments.